File: python-code/algorithms/sort/quick_sort.py
def partition(left, right, nums):
    pivot = nums[right]

    i = left - 1
    for j in range(left, right):
        if nums[j] < pivot:
            i += 1
            nums[i], nums[j] = nums[j], nums[i]

    pivot_pos = i + 1
    nums[pivot_pos], nums[right] = nums[right], nums[pivot_pos]
    return pivot_pos


def quick_sort_recursive(left, right, nums):
    if left >= right:
        return

    pivot_pos = partition(left, right, nums)

    quick_sort_recursive(left, pivot_pos - 1, nums)
    quick_sort_recursive(pivot_pos + 1, right, nums)


def quick_sort(nums):
    quick_sort_recursive(0, len(nums) - 1, nums)


nums = [5, 2, 7, 3, 8, 3, 4, 5, 6, 8, 2, 4]
quick_sort(nums)

File: java-code/src/main/java/me/joohyuk/sort/QuickSort.java
package me.joohyuk.sort;

public class QuickSort {
    private static int partition(int left, int right, int[] nums) {

        int pivotNum = nums[right];
        int j = left - 1;

        for (int i = left; i < nums.length; i++) {
            if (nums[i] < pivotNum) {
                j += 1;

                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
            }
        }

        int temp = nums[j + 1];
        nums[j + 1] = nums[right];
        nums[right] = temp;
        return j + 1;
    }

    private static void sort(int left, int right, int[] nums) {
        if (left >= right) {
            return;
        }

        int pivot = partition(left, right, nums);

        sort(left, pivot - 1, nums);
        sort(pivot + 1, right, nums);
    }

    private static void quickSort(int[] nums) {
        sort(0, nums.length - 1, nums);
    }

    public static void main(String[] args) {
        int[] nums = {5, 2, 3, 4, 65, 6, 45, 3, 657, 456, 7, 456, 7};
        quickSort(nums);
        for (int num : nums) {
            System.out.print(num + " ");
        }
    }
}
